zerver/lib/streams.py

```diff
@@ -1118,52 +1128,70 @@ def filter_stream_authorization(
-        # Deactivated streams are not accessible
-        if stream.deactivated:
-            unauthorized_streams.append(stream)
-            continue
+        is_subscribed = stream.recipient_id in subscribed_recipient_ids
+        if user_has_content_access(
+            user_profile,
+            stream,
+            user_group_membership_details,
+            is_subscribed=is_subscribed,
+            exclude_zephyr_realms_from_public_check=exclude_zephyr_realms_from_public_check,
+        ):
+            content_access_streams.append(stream)
 
-        # The user is authorized for their own streams
-        if stream.recipient_id in subscribed_recipient_ids:
-            continue
+    return content_access_streams
 
-        # Web-public streams are accessible even to guests
-        if stream.is_web_public:
-            continue
 
-        if user_profile.is_guest:
-            unauthorized_streams.append(stream)
-            continue
+def filter_stream_authorization_for_adding_subscribers(
+    user_profile: UserProfile, streams: Collection[Stream], is_subscribing_other_users: bool = False
+) -> StreamsCategorizedByPermissionsForAddingSubscribers:
+    if len(streams) == 0:
+        return StreamsCategorizedByPermissionsForAddingSubscribers(
+            authorized_streams=[],
+            unauthorized_streams=[],
+            streams_to_which_user_cannot_add_subscribers=[],
+        )
 
-        # Members and administrators are authorized for public streams
-        if not stream.invite_only:
-            continue
+    # For adding subscribers, we consider streams in zephyr realm
+    # public.
+    user_group_membership_details = UserGroupMembershipDetails(user_recursive_group_ids=None)
+    content_access_streams = get_content_access_streams(
+        user_profile,
+        streams,
+        user_group_membership_details,
+        exclude_zephyr_realms_from_public_check=True,
+    )
+    content_access_stream_ids = {stream.id for stream in content_access_streams}
 
-        if stream.invite_only:
-            user_recursive_group_ids = set(
-                get_recursive_membership_groups(user_profile).values_list("id", flat=True)
+    streams_to_which_user_cannot_add_subscribers: list[Stream] = []
+    if is_subscribing_other_users:
+        streams_to_which_user_cannot_add_subscribers = (
+            get_streams_to_which_user_cannot_add_subscribers(
+                content_access_streams,
+                user_profile,
+                user_group_membership_details=user_group_membership_details,
             )
-            # The above has checked that the user is not a guest for the below settings.
-            if is_user_in_groups_granting_content_access(stream, user_recursive_group_ids):
-                continue
+        )
 
-        unauthorized_streams.append(stream)
+    unauthorized_streams = [
+        stream
+        for stream in streams
+        if stream.deactivated or stream.id not in content_access_stream_ids
```

Comment:
Probably this `stream.deactivated` check is wrong, but another PR is responsible for removing it; I think @shuklamaneesh23's PR for #33457 is what should the `deactivated` part of the check here.

Reply:
I don't think this should be removed because we do not want to allow adding subscribers to archived channels.

Reply:
I am not sure about that. Let me start a design discussion.